File: app/src/main/java/com/sg/moviesindex/adapter/ReviewsAdapter.java
package com.sg.moviesindex.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.sg.moviesindex.R;
import com.sg.moviesindex.databinding.ReviewListItemBinding;
import com.sg.moviesindex.model.tmdb.Review;
import com.sg.moviesindex.model.tmdb.ReviewsList;

public class ReviewsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final ReviewsList reviewsList;

    public ReviewsAdapter(ReviewsList reviewsList) {
        this.reviewsList = reviewsList;
    }


    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        ReviewListItemBinding reviewListViewBinding = DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.review_list_item, viewGroup, false);
        return new ReviewsViewHolder(reviewListViewBinding);

    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int i) {
        if (viewHolder instanceof ReviewsViewHolder) {
            Review review = reviewsList.getResults().get(i);
            ((ReviewsViewHolder) viewHolder).reviewListViewBinding.setReview(review);
        }
    }

    static class ReviewsViewHolder extends RecyclerView.ViewHolder {
        private final ReviewListItemBinding reviewListViewBinding;

        ReviewsViewHolder(@NonNull final ReviewListItemBinding reviewListViewBinding) {
            super(reviewListViewBinding.getRoot());
            this.reviewListViewBinding = reviewListViewBinding;
        }
    }

    @Override
    public int getItemCount() {
        return reviewsList == null || reviewsList.getResults() == null ? 0 : reviewsList.getResults().size();
    }

}

File: app/src/main/java/com/sg/moviesindex/ui/details/ReviewsAdapter.kt
package com.sg.moviesindex.ui.details

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.sg.moviesindex.R
import com.sg.moviesindex.data.remote.Review
import com.sg.moviesindex.databinding.ReviewListItemBinding

class ReviewsAdapter : ListAdapter<Review, RecyclerView.ViewHolder>(DiffCallback()) {
  override fun onCreateViewHolder(
    viewGroup: ViewGroup,
    viewType: Int,
  ): RecyclerView.ViewHolder {
    val reviewListViewBinding: ReviewListItemBinding =
      DataBindingUtil.inflate(
        LayoutInflater.from(viewGroup.context),
        R.layout.review_list_item,
        viewGroup,
        false,
      )
    return ReviewsViewHolder(reviewListViewBinding)
  }

  override fun onBindViewHolder(
    viewHolder: RecyclerView.ViewHolder,
    i: Int,
  ) {
    if (viewHolder is ReviewsViewHolder) {
      val review = getItem(i)
      if (review != null) {
        viewHolder.reviewListViewBinding.review = review
      }
    }
  }

  class ReviewsViewHolder(
    val reviewListViewBinding: ReviewListItemBinding,
  ) : RecyclerView.ViewHolder(reviewListViewBinding.root)

  class DiffCallback : DiffUtil.ItemCallback<Review>() {
    override fun areItemsTheSame(
      oldItem: Review,
      newItem: Review,
    ): Boolean = oldItem.id == newItem.id

    override fun areContentsTheSame(
      oldItem: Review,
      newItem: Review,
    ): Boolean = oldItem == newItem
  }
}
